cpp/src/arrow/engine/substrait/extension_set.h

```diff
@@ -0,0 +1,137 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// This API is EXPERIMENTAL.
+
+#pragma once
+
+#include <vector>
+
+#include "arrow/engine/visibility.h"
+#include "arrow/type_fwd.h"
+#include "arrow/util/optional.h"
+#include "arrow/util/string_view.h"
+
+namespace arrow {
+namespace engine {
+
+/// A mapping from arrow types and functions to the (uri, name) which identifies
+/// the corresponding substrait extension. Substrait types and variations must be
+/// registered with their corresponding arrow::DataType before they can be used!
+class ARROW_ENGINE_EXPORT ExtensionIdRegistry {
+ public:
+  /// All uris registered in this ExtensionIdRegistry
```

Comment:
I don't think you're slow on the uptake, because I've been looking at this code for over an hour now and I'm still completely clueless about how ownership works here and what references what. This part of @bkietz's code is simply above my C++ skill level.

If you're sure of your understanding of this code, can you please suggest changes that you think clarify the code via github's UI or via a PR into this branch so I can easily merge them in? I don't feel confident in my ability to figure out what needs to be changed otherwise.

Reply:
I took a pass at simplifying this here (https://github.com/jvanstraten/arrow/pull/1/commits/8bd00a26cff8c2dcc7736fc771e7c739f9990755).  I understand now pretty well what is going on.  I don't think the un-simplified version is wrong so this is a purely optional commit.  If you would prefer I can open this as a PR and we can wait for @bkietz to return.

There is one small change we do want to include either way.  I mentioned it here: https://github.com/jvanstraten/arrow/pull/1/commits/8bd00a26cff8c2dcc7736fc771e7c739f9990755#r802239777